llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -5342,6 +5342,10 @@ LoopVectorizationCostModel::getUniformMemOpCost(Instruction *I,
   StoreInst *SI = cast<StoreInst>(I);
 
   bool IsLoopInvariantStoreValue = Legal->isInvariant(SI->getValueOperand());
+  // TODO: We have tests that request the cost of extracting element
+  // VF.getKnownMinValue() - 1 from a scalable vector. This is actually
+  // meaningless, given what we actually want is the last lane and is likely
+  // to be more expensive.
```

Comment:
The issue is generally that we allow to invariant addresses with scalable vectors and for those we may have to extract the last lane, but we query the last *known* lane, which is not what we need for scalable vectors, right?

Might be good to adjust the comment a bit, we have tests to exercise this code, but might be clearer to just explain the problematic code path?

Also, I presume that there's no way to query TTI for the last lane of a scalable vector at the moment?

Reply:
Maybe we could change the Index argument of `getVectorInstrCost` to be an ElementCount?

Reply:
So I'm looking at changing the `getVectorInstrCost` interface to add support for other "unknown" indices that are more meaningful, such as the last element of a vector. In the meantime, I'm happy to update the comment. There are already quite a few tests exercising this code path (since we make different vectorisation decisions with my downstream patch). Once I've got something working I'll put up a WIP patch for people to look at.

Reply:
I see, FWIW `VPLane::Kind` has something similar to represent the last lane:

```c++
  /// Kind describes how to interpret Lane.
  enum class Kind : uint8_t {
    /// For First, Lane is the index into the first N elements of a
    /// fixed-vector <N x <ElTy>> or a scalable vector <vscale x N x <ElTy>>.
    First,
    /// For ScalableLast, Lane is the offset from the start of the last
    /// N-element subvector in a scalable vector <vscale x N x <ElTy>>. For
    /// example, a Lane of 0 corresponds to lane `(vscale - 1) * N`, a Lane of
    /// 1 corresponds to `((vscale - 1) * N) + 1`, etc.
    ScalableLast
  };
```

Reply:
> Might be good to adjust the comment a bit, we have tests to exercise this code, but might be clearer to just explain the problematic code path?

Looks like the comment update may have been missed?

Reply:
Hi @fhahn @lukel97 I've created a PR to add support for querying the cost of extracting the last lane in a scalable vector. I'll add reviewers (and split up the PR if necessary) if you think it looks good! https://github.com/llvm/llvm-project/pull/144086